base_layer/wallet_ffi/src/lib.rs

```diff
@@ -1785,6 +1786,125 @@ pub unsafe extern "C" fn wallet_create(
+///
+/// ## Arguments
+/// `wallet` - The TariWallet pointer.
+/// `msg` - The message pointer.
+/// `error_out` - Pointer to an int which will be modified to an error code should one occur, may not be null. Functions
+/// as an out parameter.
+/// ## Returns
+/// `*mut c_char` - Returns the pointer to the hexadecimal representation of the signature and
+/// public nonce, seperated by a pipe character. Empty if an error occured.
+#[no_mangle]
+pub unsafe extern "C" fn wallet_sign_message(
+    wallet: *mut TariWallet,
+    msg: *mut c_char,
+    error_out: *mut c_int,
+) -> *mut c_char
+{
+    let mut error = 0;
+    ptr::swap(error_out, &mut error as *mut c_int);
+
+    let mut result = CString::new("").unwrap();
+    let nonce;
+    let message;
+    let secret;
+    if msg.is_null() {
+        error = LibWalletError::from(InterfaceError::NullError("message".to_string())).code;
+        ptr::swap(error_out, &mut error as *mut c_int);
+    } else {
+        nonce = TariPrivateKey::random(&mut OsRng);
+        secret = (*wallet).comms.node_identity().secret_key().clone();
+        message = CString::from_raw(msg).to_str().unwrap().to_owned();
+        let challenge = Blake256::digest(message.as_bytes());
+        let signature = RistrettoSchnorr::sign(secret, nonce, challenge.as_slice());
+        match signature {
+            Ok(s) => {
+                let hex_sig = s.get_signature().to_hex();
+                let hex_nonce = s.get_public_nonce().to_hex();
+                let hex_return = format!("{}|{}", hex_sig, hex_nonce);
+                result = CString::new(hex_return).unwrap();
+            },
+            Err(e) => {
+                error = LibWalletError::from(e).code;
+                ptr::swap(error_out, &mut error as *mut c_int);
+            },
+        }
+    }
+    result.into_raw()
+}
+
+/// Verifies the signature of the message signed by a TariWallet
+///
+/// ## Arguments
+/// `wallet` - The TariWallet pointer.
+/// `hex_sig_nonce` - The pointer to the sting containing the hexadecimal representation of the
+/// signature and public nonce seperated by a pipe character.
+/// `hex_sig_nonce` - The pointer to the msg the signature will be checked against.
+/// `error_out` - Pointer to an int which will be modified to an error code should one occur, may not be null. Functions
+/// as an out parameter.
+/// ## Returns
+/// `bool` - Returns if the signature is valid or not, will be false if an error occurs.
+#[no_mangle]
+pub unsafe extern "C" fn wallet_verify_message_signature(
+    public_key: *mut TariPublicKey,
+    hex_sig_nonce: *mut c_char,
+    msg: *mut c_char,
+    error_out: *mut c_int,
+) -> bool
+{
+    let mut error = 0;
+    ptr::swap(error_out, &mut error as *mut c_int);
+
+    let mut result = false;
+    if public_key.is_null() {
+        error = LibWalletError::from(InterfaceError::NullError("public key".to_string())).code;
+        ptr::swap(error_out, &mut error as *mut c_int);
+    } else {
```

Comment:
Instead of all these nested if-then-elses, it's easier to read when you return early, so
```
if pubkey.is_null() {
  ...
  return result;
}
if err_cond() {
    ...
    return result;
}
// everything is good
...
result
```

https://blog.timoxley.com/post/47041269194/avoid-else-return-early

Reply:
Done